### PR TITLE
Upgrade libp2p to 0.49.0

### PR DESCRIPTION
**Pull request by dmitry-markin**
Resolves https://github.com/paritytech/substrate/issues/12267 and pulls in https://github.com/libp2p/rust-libp2p/pull/2846 to support fallback Kademlia protocol names (see https://github.com/paritytech/substrate/issues/7746).

The main change is that `NetworkBehaviourEventProcess` was deprecated, so the event processing previously taking place in `Behaviour` popped up to `NetworkWorker`.

### Follow-Up Issue

Transports and protocols currently used rely on `async-std` primitives, what can cause issues under the `tokio` runtime. They should be upgraded to use `tokio` implementations: https://github.com/paritytech/substrate/issues/12499.

### CODE REVIEW

**Review by melekes — COMMENTED** [review]
nice 👍 what's blocking you from finishing this?

**Comment by gilescope** [comment]
This upgrade also removes an unmaintained dependency from being pulled in (a few levels down) so I would love to see this landed.

**Comment by dmitry-markin** [comment]
> nice +1 what's blocking you from finishing this?

@melekes There is a blocking bug in libp2p-0.48 with `NetworkBehaviour` derive macro that was fixed in master and is going to be released with 0.49. So we are waiting for the next release of libp2p.

**Comment by gilescope** [comment]
Once we've done this PR, we should do another PR to bump kvdb-rocksdb to 0.16 (like https://github.com/paritytech/substrate/pull/11144/files ) as the current version is also dependent on the unmaintained crate. That second PR should reduce the pulling in of multiple crate versions that this PR will create. (feel free to ping me to do it if you don't want to, just want to make sure it gets done.)

**Comment by jasl** [comment]
+1 
Hope these dependencies bumping can be included in coming `polkadot-v0.9.30`

**Comment by dmitry-markin** [comment]
Pulled the changes from master in, but still waiting for libp2p-0.49.0 to be released...

**Review by bkchr — APPROVED** [review]

**Review by melekes — APPROVED** [review]
👍

File: client/network/src/behaviour.rs [hunk 1]
```diff
@@ -186,15 +164,60 @@ pub enum BehaviourOut<B: BlockT> {
...
+		pending_response: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+	},
+
+	/// A new warp sync request must be emitted.
+	WarpSyncRequest {
+		/// Node we send the request to.
+		target: PeerId,
+		/// Warp sync request.
+		request: WarpProofRequest<B>,
+		/// One-shot channel to recieve the response.
```
Comment (altonen) [inline client/network/src/behaviour.rs:193]:
```suggestion
		/// One-shot channel to receive the response.
```

File: client/network/src/behaviour.rs [hunk 2]
```diff
@@ -186,15 +164,60 @@ pub enum BehaviourOut<B: BlockT> {
 		messages: Vec<(ProtocolName, Bytes)>,
 	},
 
+	/// A new block request must be emitted.
+	BlockRequest {
+		/// Node we send the request to.
+		target: PeerId,
+		/// Opaque implementation-specific block request.
+		request: OpaqueBlockRequest,
+		/// One-shot channel to recieve the response.
```
Comment (altonen) [inline client/network/src/behaviour.rs:173]:
```suggestion
		/// One-shot channel to receive the response.
```

File: client/network/src/behaviour.rs [hunk 3]
```diff
@@ -186,15 +164,60 @@ pub enum BehaviourOut<B: BlockT> {
...
+		pending_response: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+	},
+
+	/// A new state request must be emitted.
+	StateRequest {
+		/// Node we send the request to.
+		target: PeerId,
+		/// Opaque implementation-specific state request.
+		request: OpaqueStateRequest,
+		/// One-shot channel to recieve the response.
```
Comment (altonen) [inline client/network/src/behaviour.rs:183]:
```suggestion
		/// One-shot channel to receive the response.
```

**Review by altonen — APPROVED** [review]

**Comment by dmitry-markin** [comment]
bot merge

**Pull request merged**